File: solutions/clojure/squeaky-clean/src/squeaky_clean.clj
(ns squeaky-clean
  (:require [clojure.string :as str]))

(defn clean
  "a partial set of routines to help a developer clean up identifier names"
  [input]
  (-> input
      ;; Replace any spaces encountered with underscores
      (str/replace #"\s" "_")
      ;; Replace control characters with the upper case string "CTRL"
      (str/replace #"\p{Cntrl}" "CTRL")
      ;; Convert kebab-case to camelCase
      (str/replace #"(-\p{L})" #(str/upper-case (%1 1)))
      ;; Omit characters that are not letters
      (str/replace #"[^\p{L}_]" "")
      ;; Omit Greek lower case letters
      (str/replace #"[α-ω]" "")))

File: solutions/fsharp/squeaky-clean/SqueakyClean.fs
module SqueakyClean

open System
open System.Text.RegularExpressions

let transform (c: char) : string = 
    match c with
    | '-' -> "_"                                      // Replace any hyphens encountered with underscores
    | c when Char.IsWhiteSpace c -> ""                // Remove all whitespace
    | c when Char.IsUpper c -> $"-{Char.ToLower c}"   // Convert camelCase to kebab-case
    | c when Char.IsNumber c -> ""                    // Omit characters that are digits
    | c when Regex.IsMatch(string c, @"[α-ω]") -> "?" // Replace Greek lower case letters with question marks
    | _ -> string c
    
let clean (identifier: string): string =
    identifier |> String.collect (fun (c: char) -> transform c)
